src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -759,3 +759,248 @@ - (void)invokeCommandWithEndpointID:(NSNumber *)endpointID
+            return NO;
+        }
+    }
+    return YES;
+}
+
+static BOOL MTRDictionaryIsStructurallyValid(NSDictionary * input, NSDictionary * structure)
+{
+    for (id key in structure) {
+        id inputValue = input[key];
+        if (!inputValue) {
+            MTR_LOG_ERROR("Input does not have key %@: %@, %@", key, input, structure);
+            return NO;
+        }
+        id valueStructure = structure[key];
+        if (!MTRInputIsStructurallyValid(inputValue, valueStructure)) {
+            MTR_LOG_ERROR("Dictionary value not structurally valid: %@, %@", inputValue, valueStructure);
+            return NO;
+        }
+    }
+
+    return YES;
+}
+
+static BOOL MTRAttributeReportIsStructurallyValid(id input)
+{
+    // Input is an array of values, but they could have slightly different
+    // structures, so we can't do a single MTRInputIsStructurallyValid to check
+    // that.  Check the items one at a time.
+    if (![input isKindOfClass:NSArray.class]) {
+        MTR_LOG_ERROR("Attribute report is not an array: %@", input);
+        return NO;
+    }
+
+    NSArray * inputArray = input;
+    for (id item in inputArray) {
+        // item can be a value report or an error report.
+        if (!MTRInputIsStructurallyValid(item, @ {
+                MTRAttributePathKey : MTRAttributePath.class,
+                MTRDataKey : MTRInputStructureDataValue,
+            })
+            && !MTRInputIsStructurallyValid(item, @ {
+                   MTRAttributePathKey : MTRAttributePath.class,
+                   MTRErrorKey : NSError.class,
+               })) {
+            MTR_LOG_ERROR("Attribute report contains a weird entry: %@", item);
+            return NO;
+        }
+
+        // Now we know item is in fact a dictionary, and it has at least one of MTRDataKey and MTRErrorKey.  Make sure it's
+        // not claiming both, which could confuse code that examines it.
+        NSDictionary * itemDictionary = item;
+        if (itemDictionary[MTRDataKey] != nil && itemDictionary[MTRErrorKey] != nil) {
+            MTR_LOG_ERROR("Attribute report contains an entry that claims to be both data and error: %@", item);
+            return NO;
+        }
+    }
+
+    return YES;
+}
+
+static BOOL MTREventReportIsStructurallyValid(id input)
+{
+    // Input is an array of values, but they could have slightly different
+    // structures, so we can't do a single MTRInputIsStructurallyValid to check
+    // that.  Check the items one at a time.
+    if (![input isKindOfClass:NSArray.class]) {
+        MTR_LOG_ERROR("Event report is not an array: %@", input);
+        return NO;
+    }
+
+    NSArray * inputArray = input;
+    for (id item in inputArray) {
+        // item can be a value report or an error report.
+        // MTREventIsHistoricalKey is claimed to be present no matter what, as
+        // long as MTREventPathKey is present.
+        if (!MTRInputIsStructurallyValid(item, @ {
```

Comment:
It would be good to create the constant "structure" objects once (e.g. with a dispatch_once) rather than each time we validate something. I'm pretty the class reference don't get constant-folder by the compiler.

Reply:
Changed this to use a static for the "structure".